openpgp/signature: support for Policy URI subpackets

jandd opened the pull request:
This commit adds support for the Policy URI subpacket type from RFC 4880
5.2.3.20.

I had previously sent this PR to golang/crypto but it has been closed
because golang/x/crypto/openpgp has been frozen.

twiss commented on openpgp/packet/signature.go:
```diff
@@ -63,7 +63,7 @@ type Signature struct {
 	SigLifetimeSecs, KeyLifetimeSecs                        *uint32
 	PreferredSymmetric, PreferredHash, PreferredCompression []uint8
-	PreferredAEAD                                           []uint8
+	PolicyURI, PreferredAEAD                                []uint8
```
This is quite nitpicky, but `PreferredAEAD` is related to the other three in the line above, while this one isn't. Could you put it on a separate line?
Also, `[]uint8` doesn't really make sense as a type here; while it's technically the same thing, the others are actual lists of numbers, this one isn't. It should probably be `string`, or at least `[]byte`.

twiss commented on openpgp/packet/signature.go:
```diff
@@ -935,7 +947,7 @@ func (sig *Signature) AddMetadataToHashSuffix() {
 	l := uint64(
 		uint64(n[0])<<56 | uint64(n[1])<<48 | uint64(n[2])<<40 | uint64(n[3])<<32 |
-		uint64(n[4])<<24 | uint64(n[5])<<16 | uint64(n[6])<<8  | uint64(n[7]))
+			uint64(n[4])<<24 | uint64(n[5])<<16 | uint64(n[6])<<8 | uint64(n[7]))
```
I know this is an autoformatter at work but could you keep the formatting as it was?

twiss commented on openpgp/packet/signature_test.go:
```diff
@@ -76,4 +76,57 @@ func TestSignUserId(t *testing.T) {
+	sig := &Signature{
+		SigType:    SigTypeGenericCert,
+		PubKeyAlgo: PubKeyAlgoRSA,
+		Hash:       0, // invalid hash function
```
Same thing here as in https://github.com/ProtonMail/go-crypto/pull/83#discussion_r662489631 :slightly_smiling_face:

jandd commented:
@twiss thanks for the review, I addressed your review comments in 89f248a

twiss commented on openpgp/packet/signature.go:
```diff
@@ -64,6 +64,7 @@ type Signature struct {
 	SigLifetimeSecs, KeyLifetimeSecs                        *uint32
 	PreferredSymmetric, PreferredHash, PreferredCompression []uint8
 	PreferredAEAD                                           []uint8
+	PolicyURI                                               []byte
```
I know I mentioned this as an option previously, but do you have a strong opinion why this shouldn't be `string`? I think it would make more sense.

twiss commented on openpgp/packet/signature.go:
```diff
@@ -416,6 +418,13 @@ func parseSignatureSubpacket(sig *Signature, subpacket []byte, isHashed bool) (r
+			return
+		}
+		sig.PolicyURI = make([]byte, len(subpacket))
+		copy(subpacket, sig.PolicyURI)
```
This `copy` is the wrong way around, so it doesn't actually currently work :sweat_smile:
(I think with a string, this would be slightly more fool-proof, see `RevocationReasonText` for inspiration.)

jandd replied:
done

twiss replied:
Thanks! Looks good, just one last nitpick; this copy is now no longer necessary.

twiss commented on openpgp/packet/signature_test.go:
```diff
@@ -76,4 +76,56 @@ func TestSignUserId(t *testing.T) {
+
+	packet, _ = Read(bytes.NewReader(buf.Bytes()))
+	sig = packet.(*Signature)
+	if sig.PolicyURI == nil || bytes.Equal(sig.PolicyURI, testPolicy) {
```
It turns out this `bytes.Equal` is missing a `!`, which hid the fact that it's broken.
(With a string, you could just use `!=`.)

jandd commented:
@twiss you are right, using a string makes the whole implementation a bit simpler. I moved PolicyURI and added a reference to the RFC section, I changed the type to string and adapted the test after merging the current code from master to avoid merge conflicts for this PR. Thanks again for your helpful review.

jandd commented:
@twiss I removed the unnecessary copy

twiss commented:
Thanks :+1: And thanks for the PR!
